Create key crashes after finish when executed from SelectSignKeyIdActivity

Create key crashes after finish when executed from SelectSignKeyIdActivity

@dschuermann 

This is quite long but please read.

After hours and hours of debugging I've found the root cause this issue. This crash is actually caused due to a NullPointerException.

     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'org.sufficientlysecure.keychain.operations.results.OperationResult$LogEntryParcel org.sufficientlysecure.keychain.operations.results.OperationResult$OperationLog.getLast()' on a null object reference
            at org.sufficientlysecure.keychain.operations.results.OperationResult.createNotify(OperationResult.java:253)
            at org.sufficientlysecure.keychain.remote.ui.SelectSignKeyIdActivity.onActivityResult(SelectSignKeyIdActivity.java:123)

From what I've understood this is the flow of this process, 
* SelectSignKeyIdActivity
* CreateKeyActivity
* All CreateKeyFragments
* CreateKeyFinalFragment
* KeychainIntentService
* Executes a EditKeyOperation
* Produces a EditKeyResult
* EditKeyResult is parceled and sent to
* CreateKeyFinalFragment and to
* SelectSignKeyIdActivity
* EditKeyResult is unparceled(recreated using the static CREATOR)

While parceling the mLog object in the OperationResult(parent of EditKeyResult) is dehydrated(put into a static Map) and while unparceling the object is rehydrated(removed from the static Map).

This is where the problem is, since SelectSignKeyIdActivity is a Remote API class(usually called from a different ClassLoader), this is having access only to another static instance of the Map which is always empty. So the rehydrate() function always returns null.

This is the debug output after creating two keys from SelectSignKeyIdActivity. This shows the dehydrateMap after parceling and unparceling.

     03-15 09:32:23.081  15657-15657/org.sufficientlysecure.keychain E/Keychain? {49cac3a6-4069-4446-b3a5-f1fc17a2c7b0=org.sufficientlysecure.keychain.operations.results.OperationResult$OperationLog@c857b8}
     03-15 09:32:23.111  15487-15487/org.sufficientlysecure.keychain:remote_api E/Keychain? {}
     
     03-15 09:33:28.491  15657-15657/org.sufficientlysecure.keychain E/Keychain? {49cac3a6-4069-4446-b3a5-f1fc17a2c7b0=org.sufficientlysecure.keychain.operations.results.OperationResult$OperationLog@c857b8, fde63cec-5a89-4e05-aa5d-80f16ad20bc3=org.sufficientlysecure.keychain.operations.results.OperationResult$OperationLog@361a0b6f}
     03-15 09:33:28.521  15487-15487/org.sufficientlysecure.keychain:remote_api E/Keychain? {}

If you see in the second set of output, the log object created in the previous call is still there in the Map.

I also assume this issue exist with other Remote API classes too.

--- a/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/ui/AccountSettingsFragment.java
+++ b/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/ui/AccountSettingsFragment.java
@@ -28,8 +28,8 @@ import android.widget.TextView;
 
 import org.sufficientlysecure.keychain.Constants;
 import org.sufficientlysecure.keychain.R;
+import org.sufficientlysecure.keychain.operations.results.EditKeyResult;
 import org.sufficientlysecure.keychain.operations.results.OperationResult;
-import org.sufficientlysecure.keychain.operations.results.SaveKeyringResult;
 import org.sufficientlysecure.keychain.pgp.KeyRing;
 import org.sufficientlysecure.keychain.remote.AccountSettings;
 import org.sufficientlysecure.keychain.ui.CreateKeyActivity;
@@ -106,8 +106,8 @@ public class AccountSettingsFragment extends Fragment {
             case REQUEST_CODE_CREATE_KEY: {
                 if (resultCode == Activity.RESULT_OK) {
                     if (data != null && data.hasExtra(OperationResult.EXTRA_RESULT)) {
-                        SaveKeyringResult result = data.getParcelableExtra(OperationResult.EXTRA_RESULT);
-                        mSelectKeySpinner.setSelectedKeyId(result.mRingMasterKeyId);
+                        EditKeyResult result = data.getParcelableExtra(OperationResult.EXTRA_RESULT);
+                        mSelectKeySpinner.setSelectedKeyId(result.mMasterKeyId);
                     } else {
                         Log.e(Constants.TAG, "missing result!");
                     }

--- a/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/ui/SelectSignKeyIdActivity.java
+++ b/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/ui/SelectSignKeyIdActivity.java
@@ -127,9 +127,9 @@ public class SelectSignKeyIdActivity extends BaseActivity {
             case REQUEST_CODE_CREATE_KEY: {
                 if (resultCode == Activity.RESULT_OK) {
                     if (data != null && data.hasExtra(OperationResult.EXTRA_RESULT)) {
-//                        SaveKeyringResult result = data.getParcelableExtra(OperationResult.EXTRA_RESULT);
                         // TODO: select?
-//                            mSelectKeySpinner.setSelectedKeyId(result.mRingMasterKeyId);
+//                        EditKeyResult result = data.getParcelableExtra(OperationResult.EXTRA_RESULT);
+//                        mSelectKeySpinner.setSelectedKeyId(result.mMasterKeyId);
                     } else {
                         Log.e(Constants.TAG, "missing result!");
                     }

